src/contract.rs

```diff
@@ -0,0 +1,314 @@
+use escrow_io::*;
+use ft_main_io::*;
+use gstd::{async_main, exec, msg, prelude::*, ActorId, Encode};
+
+/// Transfers `amount` tokens from `sender` account to `recipient` account.
+/// Arguments:
+/// * `transaction_id`: associated transaction id
+/// * `from`: sender account
+/// * `to`: recipient account
+/// * `amount`: amount of tokens
+async fn transfer_tokens(
+    transaction_id: u64,
+    token_address: &ActorId,
+    from: &ActorId,
+    to: &ActorId,
+    amount_tokens: u128,
+) -> Result<(), ()> {
+    let reply = msg::send_for_reply_as::<_, FTokenEvent>(
+        *token_address,
+        FTokenAction::Message {
+            transaction_id,
+            payload: ft_logic_io::Action::Transfer {
+                sender: *from,
+                recipient: *to,
+                amount: amount_tokens,
+            }
+            .encode(),
+        },
+        0,
+    )
+    .expect("Error in sending a message `FTokenAction::Message`")
+    .await;
+
+    match reply {
+        Ok(FTokenEvent::Ok) => Ok(()),
+        _ => Err(()),
+    }
+}
+
+fn get_mut_wallet(wallets: &mut BTreeMap<WalletId, Wallet>, wallet_id: WalletId) -> &mut Wallet {
+    wallets
+        .get_mut(&wallet_id)
+        .unwrap_or_else(|| panic_wallet_not_exist(wallet_id))
+}
+
+fn reply(escrow_event: EscrowEvent) {
+    msg::reply(escrow_event, 0).expect("Error during a replying with EscrowEvent");
+}
+
+fn check_buyer_or_seller(buyer: ActorId, seller: ActorId) {
+    if msg::source() != buyer && msg::source() != seller {
+        panic!("msg::source() must be a buyer or seller");
+    }
+}
+
+fn check_buyer(buyer: ActorId) {
+    if msg::source() != buyer {
+        panic!("msg::source() must be a buyer");
+    }
+}
+
+fn check_seller(seller: ActorId) {
+    if msg::source() != seller {
+        panic!("msg::source() must be a seller");
+    }
+}
+
+fn panic_wallet_not_exist(wallet_id: WalletId) -> ! {
+    panic!("Wallet with the {wallet_id} ID doesn't exist");
+}
+
+#[async_trait::async_trait]
+pub trait EscrowHandler {
```

Comment:
What is the purpose of this trait?

Reply:
> What is the purpose of this trait?

Since our `Escrow` state defined in separate crate(`io`), we must separate concerns and implement methods via traits in contract.